Arc-d assets

andriyDev opened the pull request:
# Objective

Fixes #15723. This is an alternative to #15346 (and the associated PR #15359) obsolete. Note if we decide asset locking is preferable, there are learnings from this PR we can apply to #15359.

## Solution

1) Assets are now stored as `Arc<A>` instead of just `A`.
2) `get` remains the same, but I've added an alternative called `get_arc` which returns a clone of the Arc that stores an asset.
3) Replaced the `get_mut` with one of `get_cloned_mut`, `get_inplace_mut`, and `get_reflect_cloned_mut` (this last one is just because it's necessary for reflection, but not necessarily needed for users).
4) Added `add_arc` and `insert_arc` to allow adding `Arc` types directly.
    
    1) `add` currently takes `Into<A>`. Trying to change this to `Into<Arc<A>>` breaks a lot of callsites like `meshes.add(Rectangle::new(1.0, 2.0))` (since this requires two `Into` calls, which Rust can't figure out).
    2) As a result, making `add_arc` take `Into<Arc<A>>` is a fine middle ground to allow both `Arc`-ed assets and owned assets.
    3) For completeness, I also added `insert_arc`.

5) `RenderAsset` now takes an `Arc<Self::SourceAsset>`. This means in most cases, cloning the asset can be skipped when passing data to the render world. Data only needs to be cloned if you mutate an asset.

## Testing

- Added an example that shows off using an Arc-d asset in an async context.

---

## Showcase

Assets can now be used in asynchronous contexts! By using `Assets<A>::get_arc`, you can get a clone of the underlying `Arc` of an asset, and pass it to an async closure.

```rust
fn my_task_launcher(configs: Res<Assets<ProceduralGenerationConfig>>, selected_config: Res<SelectedConfigHandle>) {
    let config = configs.get_arc(&selected_config.0).expect("The asset is loaded");
    
    AsyncComputeTaskPool::get().spawn(move async || {
        // Do something with the `config` in a task!
        std::thread::sleep(std::time::Duration::from_secs(config.time_it_takes_to_generate));
        println!("Done generating a level!");
    });
}
```

## Migration Guide

- `Assets::get_mut` has been removed. Use `get_cloned_mut` for `Clone` assets, or `get_in_place_mut` for non-`Clone` assets (and handle the case where the asset is aliased).
- `Assets::iter_mut` now returns an iterator of `&mut Arc<A>`. Consider using `Arc::make_mut` for `Clone` assets, or `Arc::get_mut` for non-`Clone` assets (and handle the case where the asset is aliased).
- `RenderAsset`s now take an `Arc<Self::SourceAsset>`. Consider using `Arc::try_unwrap()` and falling back to a clone if the asset is aliased (`try_unwrap` returns an error). Otherwise, prefer borrowing the source asset's data instead of moving it.

andriyDev committed:
Create an example for the use of asset arcing.

andriyDev committed:
Remove all the `get_mut` functions from `Assets`.

Removing these functions makes it simpler to redesign them for when assets are Arc'd.

andriyDev committed:
Make `Assets` store assets as `Arc<A>` instead of just `A`.

andriyDev committed:
Implement `get_inplace_mut`.

Now we can mutate assets in place as long as there are no more Arcs or Weaks for that asset.

andriyDev committed:
Implement `get_cloned_mut`.

Now for assets that impl `Clone`, we can get a mutable borrow to the asset always by cloning.

andriyDev committed:
Implement `get_reflect_cloned_mut`.

This is basically the same as `get_cloned_mut` but using reflection. This is not really intended to be used publicly, but the reflect.rs needs a clone that only relies on reflection, so making it public seems fine.

andriyDev committed:
Update most calls in bevy_asset to either use a `.into()` or one of the `get_*_mut` variants.

andriyDev committed:
Make RenderAssets take an `Arc` to the source asset and extract source assets as Arcs.

andriyDev committed:
Separate out inserting/adding an owned asset from the arc version.

Many callers expect to be able to call `meshes.add(Rectangle::new(50.0,
100.0))`. Now we can keep all these callsites the same (while still
supporting inserting/adding `Arc`s directly).

andriyDev committed:
Make all the crates use the new `get_*_mut` variants, or use `Arc<SourceAsset>` in `RenderAsset`.

andriyDev committed:
Update the examples to use the `get_*_mut` variants or otherwise dealing with `Arc`s.

andriyDev committed:
Add a nice big warning on `get_arc` about holding the `Arc` for long periods of time.

BenjaminBrienen commented:
Can you fix the inconsistency between "[`Arc`]-ed" and "[`Arc`]d"?

aecsocket reviewed (commented):
I really like the concept of storing assets in an `Arc`. I find that I very rarely mutate existing assets, only add/remove/swap, so `Arc`s for assets make perfect sense to me. I am somewhat worried about a lot of usages of `get_cloned_mut` that this causes in the rest of the codebase though.

andriyDev committed:
Rewrite any "Arc-ed" to "Arc-d" to make it consistent with other uses.

andriyDev committed:
Improve the MutableAssetError messages to match Rust guidelines.

https://rust-lang.github.io/api-guidelines/interoperability.html?highlight=error#error-types-are-meaningful-and-well-behaved-c-good-err

andriyDev committed:
Rename `get_inplace_mut` to `get_in_place_mut`.

andriyDev committed:
Fix the example using bad wording.

andriyDev commented on examples/asset/arc_asset.rs:
```diff
@@ -0,0 +1,251 @@
+//! This example illustrates how to use assets in an async context (through locking).
```
I dropped the whole "locking" wording, that was actually just an artifact from my "asset locking" PR (which I stole this example from, lol).

aecsocket replied:
I still do like the naming of "locking" an asset, since that is effectively what you're doing when you've got 2+ ref counts to an asset - you're locking it for mutation. Maybe this term could be used in docs somewhere?
```
/// When an asset has a reference count of more than 1, this asset is considered *locked* for mutation.
```

andriyDev commented on crates/bevy_asset/src/assets.rs:
```diff
@@ -415,29 +420,58 @@ impl<A: Asset> Assets<A> {
+    /// returns an error.
+    ///
+    /// [`Weak`]: std::sync::Weak
+    pub fn get_inplace_mut(
```
Done. I don't have strong opinions here, this seems fine!

andriyDev commented on crates/bevy_asset/src/direct_access_ext.rs:
```diff
@@ -9,6 +11,9 @@ pub trait DirectAssetAccessExt {
     fn add_asset<A: Asset>(&mut self, asset: impl Into<A>) -> Handle<A>;
 
+    /// Insert an [`Arc`]ed asset similarly to [`Assets::add_arc`].
+    fn add_arc_asset<A: Asset>(&mut self, asset: impl Into<Arc<A>>) -> Handle<A>;
```
I think I'd prefer `Into<Arc<A>>` because it allows for other types to be used like for example `Box<A>`. It definitely could be `Arc<A>`, but I'd prefer to err on the side of caution here.

aecsocket replied:
Fair enough

andriyDev commented on crates/bevy_asset/src/assets.rs:
```diff
@@ -620,6 +708,14 @@ pub struct InvalidGenerationError {
+    #[display("The asset is not present or has an invalid generation.")]
+    Missing,
+    #[display("The asset Arc is aliased (there is another Arc or Weak to this asset), so it is not safe to mutate.")]
+    Aliased,
```
Done.

andriyDev commented on crates/bevy_asset/src/assets.rs:
```diff
@@ -509,13 +543,13 @@ impl<A: Asset> Assets<A> {
                         marker: PhantomData,
                     };
-                    (id, v)
+                    (id, v.clone())
```
Technically, there's some possible contention on the refcount when cloning, but by not cloning, we incur a double dereference whenever accessing the Arc. Generally I prefer to just clone an `Arc` since what you really care about is accessing the data, not the container. A borrow to an `Arc` is just adding another layer of indirection. Having no lifetime on the iterator item is nice!

aecsocket replied:
I didn't consider the double deref, although I think you'd still be going through the first deref anyway if you always cloned the Arc. I agree that having ownership of the Arc is nicer for the user though, so I'm happy with `Arc<A>`.

Doesn't this PR *add* a lifetime to the iterator item, with the `+ '_`?
```diff
-    pub fn iter(&self) -> impl Iterator<Item = (AssetId<A>, &A)> {
+    pub fn iter(&self) -> impl Iterator<Item = (AssetId<A>, Arc<A>)> + '_ {
```
Unless I'm misunderstanding something :sweat_smile:

andriyDev replied:
This PR "adds" a lifetime to the *iterator* not the iterator item. Note the previous implementation also had this same lifetime, it was just elided since the return type includes a borrow. So we're just making the lifetime explicit. The difference for the new implementation is that we get ownership of the Arc like you mentioned.

aecsocket replied:
Ah, the lifetime was elided because of the `&A`, makes sense. I get it now.

andriyDev commented:
> Can you fix the inconsistency between "[`Arc`]-ed" and "[`Arc`]d"? I prefer the latter.

Done!

andriyDev commented on crates/bevy_asset/src/direct_access_ext.rs:
```diff
@@ -28,6 +33,14 @@ impl DirectAssetAccessExt for World {
+    ///
+    /// # Panics
+    /// If `self` doesn't have an [`AssetServer`] resource initialized yet.
+    fn add_arc_asset<A: Asset>(&mut self, asset: impl Into<Arc<A>>) -> Handle<A> {
```
Same as above ;)

aecsocket replied:
Same as above :+1:

andriyDev commented on crates/bevy_asset/src/assets.rs:
```diff
@@ -572,15 +606,69 @@ impl<A: Asset> Assets<A> {
+            );
+        }
+
+        let cloned_asset = FromReflect::from_reflect(arc.as_ref()).expect(
+            "could not call `FromReflect::from_reflect` in `Assets::get_reflect_cloned_mut`",
+        );
```
This is the same message used in `ReflectAsset`. It's not clear to me that the message you suggested is correct. `Assets<A>` already holds `A`, so calling `FromReflect<A>` should never fail in theory. I think using a generic error message is fine here.

aecsocket replied:
I didn't realise this message was already used elsewhere, so I'm fine with the original message.

andriyDev committed:
Switch MutableAssetError to use thiserror instead of derive_more.

andriyDev commented:
Straight forward rebase to do with #16368, and updated error to using thiserror after #16684.

andriyDev committed:
Fix the `asset_changed` test by using `get_inplace_mut`.

andriyDev committed:
Make the docs for `AssetChanged` reference `get_cloned_mut` instead of the old `get_mut`.

andriyDev committed:
Fix the `mixed_lighting` example to use `get_cloned_mut`.

andriyDev committed:
Fix the `many_materials` and `edit_material_on_gltf` examples.

andriyDev committed:
Fix the `specular_tint` example.

mockersf committed:
Merge branch 'main' into arc-assets

andriyDev committed:
Replace `gen` with `r#gen` to transition to edition 2024.

andriyDev committed:
Fix the many_cubes stress test to use Arc::make_mut for the materials.

ElliottjPierce reviewed (approved):
First, this is a really good implementation. It is irreducibly complex and well thought out. I don't see any bugs or red flags, so I'm approving it (as if I have any authority lol). 

That said, I do have two concerns.

First, is it possible to not require loading the asset from a file in your example? IDK, but if you can demonstrate this without that, I think that would be nice. If it is necessary, consider renaming the file to point back to the example. That way people know what the file is for just from its name.

Second, I am still not sure if this is the best solution to the problem. As I understand it, bevy is slowly but surely moving to assets as entities, maybe even asset entities in another world. In that set-up, there may be better alternatives to the arc available in the near future. 

It also seems pretty scary to me how easy it would be to introduce bugs with this. You mentioned it earlier, but it's really easy to forget to poll an asset for changes. What if an an asset is changed while another thread is working with the now outdated asset? Now we have to double check that all of that work is still valid. What if two changes happen at roughly the same time? Are systems going to have to re-poll constantly to make sure they have the "real" asset? This gives a lot more freedom than before; it just also opens up a lot of foot guns. For example, this could be (miss?)used to do something similar to godot's https://docs.godotengine.org/en/stable/classes/class_resource.html#class-resource-method-duplicate.

Finally, there may be times where users want/need assets to be stored densely or in `Arc<RwLock>>`.  Adding Arcs directly into assets is cool, but not as flexible as them being in the asset itself as needed (the current way). So this makes assets *much* nicer to work with in async/multi world user land but potentially much harder to use when a new user just wants a simple asset.

All that said, I think this is a reasonable and well motivated solution for now, but with assets as entities, the line between scene and asset are going to get blurry. I think that will make the asset system *way* more flexible and IMO much better. When that happens, I think we might end up preferring ECS / multi-world solutions to assuming all assets need arcs.

Anyway, the code looks good to me, and I'm approving it. I'm going to defer to others about the vision here because that's really what this comes down to. If we are going to use assets as entities through multiple worlds in the near future, this seems kinda unnecessary, but if Bevy is headed in a different direction that needs async assets in most places, this seems like the best solution available, and this PR seems like an excellent implementation. That's my thoughts anyway.

andriyDev committed:
Merge branch 'main' into arc-assets

andriyDev committed:
Rename `heights.li.ron` to `arc_asset_heights.li.ron`.

andriyDev commented:
@ElliottjPierce First, technically a file is not strictly necessary, however it is *motivating*. That is, if it wasn't a file, a reader might be confused - why are we making this an asset if we could have just passed this around as an `Arc<LinearInterpolation>` in a resource? So while I could technically make this work without a file, I think it explains the concept much better: we can load assets and still pass them around in async. That said, I did rename the file to match the name of the example.

Second, I am skeptical that assets as entities will really change much in terms of async. If anything, I expect it to make it *harder* to access assets in async, since now you need to synchronize with the whole world (or the whole asset world or something). I'm not saying we shouldn't do assets-as-entities (I very much want it), but I believe it is gonna end up being mostly orthogonal.

Next, this is absolutely a risk. However, I think the splitting of `get_mut` into `get_cloned_mut` and `get_inplace_mut` (and their respective doc comments will likely give people enough info on what's going on (I hope!). The other thing to note: both `get_*_mut` versions still require `&mut Assets<A>`. So it is still not possible to have two changes happen at the same time. So at least that is not at risk.

And yes, we are losing the "dense" nature of assets. But unlike components, I would expect most assets are fairly big anyway, and so won't benefit much from cache locality. Plus, assets are accessed in a more "random-access" way anyway. I think it's pretty rare to iterate over all assets. Users can still mimic the RwLock setup by just making their asset a `RwLock` wrapper.

I don't think this is the final solution - I agree, assets-as-entities may change things. But we should still support assets in async. This is necessary for asset saving (which I want to work on next).

Thanks for your review!!

ElliottjPierce commented:
> technically a file is not strictly necessary, however it is motivating.

Makes sense. Definitely worth keeping then, good call.

> I  am skeptical that assets as entities will really change much in terms of async.

Me too. I just mean if/when we have assets as entities in their own world, and that world is referenced by both the main and render world, etc, the cloning and sharing of assets between contexts won't be much of a motivation anymore. The async stuff is still a very strong motivation, but only for a subset of asset types that need to be referenced in a task or something. And it might be easier to do that with incremental systems or tasks in the asset world anyway. I know that context is hypothetical; just thinking ahead.

> And yes, we are losing the "dense" nature of assets. 

Good points here too. You've convinced me.

> Users can still mimic the RwLock setup by just making their asset a RwLock wrapper.

I know we aren't loosing anything here. Just feels odd that we are standardizing the Arc pattern instead of leaving it up to the user whether to do `A` , `Arc<A>`, or `Arc<RwLock<A>>`. But I suppose Arc is the most common. 

> I don't think this is the final solution - I agree, assets-as-entities may change things. 

Still, that's in the future, and as long as we're ok re-evaluating this in like a year, I do agree that this is the best solution until then.

> Next, this is absolutely a risk.

I'm ok with the additional foot guns here. We just need a really good introduction of the concept. If you have label permissions, it might be worth adding needs-release-note, but IDK how Bevy does that process.

> Thanks for your review!!

You've reviewed two of mine, so I still owe you one lol. And I'm excited for that saving work. Arcs certainly are the right choice for that, and I suppose it's a reasonable assumption that all assets should be "savable". It's always confused me why game engines have great systems for scriptable objects (unity), resources (godot), etc, but never make them savable at runtime. So I'm excited to see this come to Bevy, IIUC. Thanks!

mockersf commented:
did you check the impact of this on rendering an asset heavy scene?

mockersf commented on Cargo.toml:
```diff
@@ -1694,6 +1694,17 @@ description = "Demonstrates various methods to load assets"
 category = "Assets"
 wasm = false
 
+[[example]]
+name = "arc_asset"
+path = "examples/asset/arc_asset.rs"
+doc-scrape-examples = true
+
+[package.metadata.example.arc_asset]
+name = "Arc'd Assets"
+description = "Demonstrates how to acquire Arc'd assets and use them in an async context"
+category = "Assets"
+wasm = false
+
```
the focus of the example should not be "arc_asset", it should be "async_access_to_asset" or something. arc is the technical detail, not the interesting feature

andriyDev replied:
Good point! Done.

mockersf commented:
really dislike the public api change, but not sure there's a better way to do it

andriyDev committed:
Rename `arc_asset` example to `access_asset_from_async`.

andriyDev committed:
Merge branch 'main' into arc-assets

kristoff3r commented:
I don't understand how `get_cloned_mut` is ever the right thing to do. In case it's aliased you clone the asset, make your changes, but unless you store it somewhere or re-register with the asset system won't it just be dropped again? But none of the uses in the PR seems to do that.

I haven't read the entire PR so I might have missed something.

kristoff3r commented:
> @kristoff3r `get_cloned_mut` uses `Arc::make_mut`. It takes a **mutable** reference to an arc (in this case, the arc that is stored in `Assets`). If the arc is aliases, it clones the data, makes a new arc, **and then stores that arc in the mutable reference**. This means the arc stored in `Assets` is replaced when you call `get_cloned_mut` for an aliased asset.

Ah that makes total sense, thanks for explaining.

andriyDev commented:
@mockersf Sorry I realized I didn't respond to your other question: I have not tested this on an asset heavy scene. Does Bevy have an example or stress test of this? 

It is *possible* that the extra dereferences hurt performance, but that might not be too big a deal. We already put render assets in a hashmap so cache locality might already be hurting? Unsure.

JMS55 commented:
Fyi MeshletMesh is an example of a read-only asset that I use Arc to share with the render world cheaply.

In general, I wish we could share assets between the main/render world without any copies, in the same way that bevy schedules systems across multiple threads to avoid conflicts.

alice-i-cecile commented:
>  @cart @andriyDev I'm going to cut arc'd assets PR [from the 0.16 milestone]. I know that this is overdue for a review (and generally I'm supportive of it), but I don't think it's high enough priority to ship something high risk at the last minute or delay the release candidate

Me [on Discord](https://discord.com/channels/691052431525675048/1351525668776579142/1351646384985735269)

ElliottjPierce commented:
I made a discussion post #18949 that is an alternative to this. I don't have a crazy strong opinion, but I think my proposal is worth investigating before merging this. Just in case.

ElliottjPierce commented:
Just a head's up that @andriyDev and I have come up with a followup to this PR. See [here](https://discord.com/channels/691052431525675048/749332104487108618/1367506312455258122).

I think we need to start with merging this pr, but I think in the end, we'll have a less breaking api change and a much more flexible system.

andriyDev committed:
Merge branch 'main' into arc-assets

andriyDev committed:
Fix up a handful of cases not covered by the merge.

JMS55 commented:
> If the goal of this PR is to make assets usable in async context, why not just wrap the specific read-only assets in Arc yourself? So instead of Assets<T> Someone could just use Asset<Arc<T>> instead of wrapping everything in Arc across the board.

This is what MeshletMesh does for instance.
